File: official-docs/examples/examples/python-server-wallets/lib/wallet_storage.py
"""
Wallet Storage Utilities

WARNING: FOR TESTING AND DEVELOPMENT ONLY - NOT FOR PRODUCTION USE

This file provides simple file-based storage for wallet metadata and key shares.
It is intended ONLY for local development and testing purposes.

DO NOT USE IN PRODUCTION because:
- Key shares are stored unencrypted in a local JSON file
- No access control or security measures are implemented
- No backup or recovery mechanisms

For production environments, you should:
- Use a secure key management service (AWS KMS, HashiCorp Vault, etc.)
- Encrypt key shares at rest
- Implement proper access controls and audit logging
- Follow your organization's security best practices
"""

import json
from dataclasses import dataclass, asdict, field
from pathlib import Path

WALLET_FILE = Path.cwd() / ".wallets.json"


@dataclass
class StoredKeyShare:
    key_share_id: str
    secret_share: str


@dataclass
class StoredWallet:
    address: str
    wallet_id: str
    chain: str  # "evm" or "svm"
    created_at: str
    key_shares: list[StoredKeyShare] = field(default_factory=list)


def load_wallets() -> dict[str, StoredWallet]:
    """Load all saved wallets from local storage."""
    if not WALLET_FILE.exists():
        return {}

    try:
        data = json.loads(WALLET_FILE.read_text())
        wallets = {}
        for addr, w in data.items():
            key_shares = [StoredKeyShare(**ks) for ks in w.get("key_shares", [])]
            wallets[addr] = StoredWallet(
                address=w["address"],
                wallet_id=w["wallet_id"],
                chain=w["chain"],
                created_at=w["created_at"],
                key_shares=key_shares,
            )
        return wallets
    except (json.JSONDecodeError, TypeError, KeyError) as e:
        print(f"Warning: Failed to load wallets file: {e}")
        return {}


def save_wallet(wallet: StoredWallet) -> None:
    """Save a wallet to local storage."""
    wallets = load_wallets()
    wallets[wallet.address] = wallet

    raw = {addr: asdict(w) for addr, w in wallets.items()}
    WALLET_FILE.write_text(json.dumps(raw, indent=2))
    print(f"Wallet saved to {WALLET_FILE}")


def get_wallet(address: str) -> StoredWallet | None:
    """Get a specific wallet by address."""
    wallets = load_wallets()
    return wallets.get(address)


def list_wallets() -> list[StoredWallet]:
    """List all saved wallets."""
    wallets = load_wallets()
    return list(wallets.values())


def delete_wallet(address: str) -> bool:
    """Delete a wallet from local storage."""
    wallets = load_wallets()

    if address not in wallets:
        return False

    del wallets[address]
    raw = {addr: asdict(w) for addr, w in wallets.items()}
    WALLET_FILE.write_text(json.dumps(raw, indent=2))
    print(f"Wallet {address} deleted")
    return True

File: official-docs/examples/examples/nodejs-server-wallets/src/lib/wallet-storage.ts
/**
 * Wallet Storage Utilities
 *
 * ⚠️ WARNING: FOR TESTING AND DEVELOPMENT ONLY - NOT FOR PRODUCTION USE
 *
 * This file provides simple file-based storage for wallet key shares.
 * It is intended ONLY for local development and testing purposes.
 *
 * DO NOT USE IN PRODUCTION because:
 * - Key shares are stored unencrypted in a local JSON file
 * - No access control or security measures are implemented
 * - File permissions are not managed
 * - No backup or recovery mechanisms
 *
 * For production environments, you should:
 * - Use a secure key management service (AWS KMS, HashiCorp Vault, etc.)
 * - Encrypt key shares at rest
 * - Implement proper access controls and audit logging
 * - Use a secure database with encryption
 * - Follow your organization's security best practices
 */

import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";

// Local file storage - FOR TESTING ONLY
const WALLET_FILE = join(process.cwd(), ".wallets.json");

export interface StoredWallet {
  address: string;
  externalServerKeyShares: string[];
  createdAt: string;
}

export interface WalletStorage {
  [address: string]: StoredWallet;
}

/**
 * Load all saved wallets from local storage
 * ⚠️ FOR TESTING ONLY - Use secure storage in production
 */
export function loadWallets(): WalletStorage {
  if (!existsSync(WALLET_FILE)) return {};

  try {
    const data = readFileSync(WALLET_FILE, "utf-8");
    return JSON.parse(data);
  } catch (error) {
    console.warn("Failed to load wallets file:", error);
    return {};
  }
}

/**
 * Save a wallet to local storage
 * ⚠️ FOR TESTING ONLY - Use secure storage in production
 */
export function saveWallet(wallet: StoredWallet): void {
  const wallets = loadWallets();
  wallets[wallet.address] = wallet;

  writeFileSync(WALLET_FILE, JSON.stringify(wallets, null, 2));
  console.info(`Wallet saved to ${WALLET_FILE}`);
}

/**
 * Get a specific wallet by address
 * ⚠️ FOR TESTING ONLY - Use secure storage in production
 */
export function getWallet(address: string): StoredWallet | undefined {
  const wallets = loadWallets();
  return wallets[address];
}

/**
 * List all saved wallets
 * ⚠️ FOR TESTING ONLY - Use secure storage in production
 */
export function listWallets(): StoredWallet[] {
  const wallets = loadWallets();
  return Object.values(wallets);
}

/**
 * Delete a wallet from local storage
 * ⚠️ FOR TESTING ONLY - Use secure storage in production
 */
export function deleteWallet(address: string): boolean {
  const wallets = loadWallets();

  if (!wallets[address]) return false;

  delete wallets[address];
  writeFileSync(WALLET_FILE, JSON.stringify(wallets, null, 2));
  console.info(`Wallet ${address} deleted`);
  return true;
}
